CDC with tablets

mlitvk opened the pull request:
initial implementation to support CDC in tablets-enabled keyspaces.
The design is described in https://github.com/scylladb/scylladb/issues/19276

In this PR we:
* add new internal system tables for CDC metadata and loading it into in-memory CDC metadata
* the write coordinator chooses a stream according to this metadata
* enable creating tables with CDC enabled in tablets-enabled keyspaces. tablets are allocated for the CDC table, and a stream is created per each tablet.
* on tablet resize, the topology coordinator creates a new stream set with a stream per each new tablet, and executes the stream switching algorithm.

Fixes https://github.com/scylladb/scylladb/issues/22576

nyh commented:
> initial implementation to support CDC in tablets-enabled keyspaces. The design is described in #19276

I don't see anything described there. Is it https://docs.google.com/document/d/1qO5f2q5QoN5z1-rYOQFu6tqVLD3Ha6pphXKEqbtSNiU/edit?tab=t.0 ?
Is this document up to date?
Is it somebody's personal document, or is a copy of it in our shared dev folder (https://drive.google.com/drive/u/0/folders/0B1wZ5nEzmjS2alhkbEcyZWZpTzA?resourcekey=0-ReGtOzxXMfrU_zma8tgJhQ)?

Later, we'll also need to fix the Alternator Streams to support this new CDC implementation. For that it will be very important to have a correct and relatively up-to-date design document.

mlitvk commented:
> > initial implementation to support CDC in tablets-enabled keyspaces. The design is described in #19276
> 
> I don't see anything described there. Is it https://docs.google.com/document/d/1qO5f2q5QoN5z1-rYOQFu6tqVLD3Ha6pphXKEqbtSNiU/edit?tab=t.0 ? Is this document up to date?

Yes, this is the reference design document. It is mostly up to date except the "Deciding when to change streams" part.
We discussed it and decided to make the algorithm of changing streams be part of tablet resize, instead of being asynchronous. Now whenever the topology coordinator resizes a tablet then we create new streams at the same time for each tablet.

> Is it somebody's personal document, or is a copy of it in our shared dev folder (https://drive.google.com/drive/u/0/folders/0B1wZ5nEzmjS2alhkbEcyZWZpTzA?resourcekey=0-ReGtOzxXMfrU_zma8tgJhQ)?
> 
I think it's @piodul's shared design document. I'm not sure if it's in the dev folder

> Later, we'll also need to fix the Alternator Streams to support this new CDC implementation. For that it will be very important to have a correct and relatively up-to-date design document.

nyh commented on test/cqlpy/test_cdc.py:
```diff
@@ -22,41 +22,60 @@ def wait_for_first_cdc_generation(cql, timeout):
+            while len(list(cql.execute(query))) == 0:
+                assert time.time() < deadline, "Timed out waiting for the first CDC generation"
+                time.sleep(1)
+            print("MICHAEL ok")
```
A thought: maybe you should change wait_for_first_cdc_generation() to do the old thing or the new thing instead of having it do just the old thing and then open-coding the new thing here?

mlitvk replied:
I removed this part, it wasn't needed.
with tablets, the streams are created together with the tablets, so we don't need to wait for anything

nyh commented on test/cqlpy/test_cdc.py:
```diff
@@ -93,9 +112,6 @@ def test_cdc_with_lwt_preimage(scylla_only, cql, test_keyspace):
-@pytest.mark.parametrize("test_keyspace",
-                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #16317")]), "vnodes"],
-                         indirect=True)
 def test_cdc_taken_log_name(scylla_only, cql, test_keyspace):
```
You removed the "parameterize" that ran the test twice, once for tablets and once for vnodes. I think eventually, when we no longer care about vnodes at all (or have in CI something which runs all the tests in vnodes mode) we can do this, but perhaps not just yet?

mlitvk replied:
I wasn't sure if we we want to have parameters for every test to run twice, with vnodes and tablets, or is it something the test should be oblivious about and can be controlled from outside, like we have `--tablets` in dtest
until now we needed to have the parameters so we can mark the tablets with xfail while still allowing to run it with vnodes, but this is not relevant anymore.
but I suppose I can keep it

nyh replied:
I don't know - I think we went back and forth on this question. I think - but am **not sure** - the current decision is:
1. The test suite runs with "the default" way - e.g., today with tablets in CQL and vnodes in Alternator.
2. When we know an individual feature behaves inherently differently in vnodes and tablets, we can parameterize the test to run both. In this file this is what we did - we ran each test twice. The idea is since the implementation of CDC in both modes is quite different, as long as we support both we need to test both.

I agree we don't need both tablets and vnodes in all tests. For example, `test_cdc_taken_log_name` checks some logic of the table name that has absolutely no relevance to the vnodes/tablets split, so there is no reason to run it twice. From that test you can indeed drop the parametrization.

So I'm fine with you dropping those parameterization from some of the tests, but please just consider each one individually - whether it makes sense to try it on both tablets and vnodes because the implementation of the specific feature being tested is different - or maybe for some of the tests it's just not needed.

mlitvk replied:
I think for most of these tests in test/cql and test/cqlpy we don't need to run them twice with vnodes and tablets.
they mostly check basic properties of the cdc table, or writing and verifying few rows in a way that is not dependent on how the cdc streams are constructed, and they don't read the CDC metadata
the exception is this test `test_cdc_log_entries_use_cdc_streams` that I fixed and I kept the vnodes/tablets parameters

nyh replied:
I agree about checks for the "basic properties of the cdc table", but I think that if we have any test that actually writes data to a base table and/or reads from the CDC table they can benefit from being run in the two modes. Perhaps we don't have many of those (we have many for Alternator Streams, but this is a separate issue - https://github.com/scylladb/scylladb/issues/23838). So please just consider each test separately, and I'll trust your judgement. Thanks.

nyh commented on test/cqlpy/test_describe.py:
```diff
@@ -746,9 +746,6 @@ def test_desc_udf_uda(cql, test_keyspace, scylla_only):
 # The test is marked scylla_only because it uses a Scylla-only property "cdc".
-@pytest.mark.parametrize("test_keyspace",
-                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #16317")]), "vnodes"],
-                         indirect=True)
```
Ditto, you removed the double-running here.

nyh commented on test/cqlpy/test_describe.py:
```diff
@@ -990,9 +987,6 @@ def test_table_options_quoting(cql, test_keyspace):
 # but we are attaching `ALTER TABLE <cdc log table name> WITH <all table's properties>` to description of base table
-@pytest.mark.parametrize("test_keyspace",
-                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #16317")]), "vnodes"],
-                         indirect=True)
```
You removed the double-running here too.

nyh commented on test/cqlpy/test_describe.py:
```diff
@@ -1037,9 +1031,6 @@ def test_hide_cdc_table(scylla_only, cql, test_keyspace):
 # has the expected format.
-@pytest.mark.parametrize("test_keyspace",
-                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #16317")]), "vnodes"],
-                         indirect=True)
```
And here too. I'll stop commenting about this, but you get the picture ;-)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17095/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17095/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17095/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
#### Failed Tests (40/270):
- [test_cdc_stream_compaction.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.1)
- [test_cdc_stream_compaction.release.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.10)
- [test_cdc_stream_compaction.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.2)
- [test_cdc_stream_compaction.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.3)
- [test_cdc_stream_compaction.release.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.4)
- [test_cdc_stream_compaction.release.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.5)
- [test_cdc_stream_compaction.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.6)
- [test_cdc_stream_compaction.release.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.7)
- [test_cdc_stream_compaction.release.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.8)
- [test_cdc_stream_compaction.release.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17095/testReport/junit/cluster/test_cdc_with_tablets/test_cdc_stream_compaction_release_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cdc_stream_compaction.release.9)
#### Build Details:
- Duration: 1 hr 24 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17098/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17098/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17098/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5042/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8215/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16282/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17098/)
#### Build Details:
- Duration: 7 hr 47 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17107/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17107/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17107/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5050/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8223/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16290/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17107/)
#### Build Details:
- Duration: 8 hr 3 min
- Builder: spider3.cloudius-systems.com

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/23795)

**Changed Files:**
- [architecture/tablets](http://previews.docs.scylladb.com/scylladb/23795/architecture/tablets)
- [features/cdc/cdc-log-table](http://previews.docs.scylladb.com/scylladb/23795/features/cdc/cdc-log-table)
- [features/cdc/cdc-querying-streams](http://previews.docs.scylladb.com/scylladb/23795/features/cdc/cdc-querying-streams)
- [features/cdc/cdc-stream-changes](http://previews.docs.scylladb.com/scylladb/23795/features/cdc/cdc-stream-changes)
- [features/cdc/cdc-streams](http://previews.docs.scylladb.com/scylladb/23795/features/cdc/cdc-streams)
- [features/cdc/index.html](http://previews.docs.scylladb.com/scylladb/23795/features/cdc/index.html)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17575/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17575/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17575/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5433/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16684/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8601/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17575/)
#### Build Details:
- Duration: 9 hr 12 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17596/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17596/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17596/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16701/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8618/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5450/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17596/)
#### Build Details:
- Duration: 8 hr 20 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17625/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17625/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17625/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5479/)
:x: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16733/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8650/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17625/)
#### Failed Tests (43/20227):
- [test_create_snapshot_with_native_type_with_base_rows_delete_preimage_postimage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cdc_snapshot_operation_test/TestCDCSnapshotOperation/Tests___dtest_with_gossip_topology_changes___test_create_snapshot_with_native_type_with_base_rows_delete_preimage_postimage) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_create_snapshot_with_native_type_with_base_rows_delete_preimage_postimage)
- [test_create_snapshot_with_native_type_with_base_rows_delete_preimage_postimage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cdc_snapshot_operation_test/TestCDCSnapshotOperation/Tests___dtest_with_gossip_topology_changes___test_create_snapshot_with_native_type_with_base_rows_delete_preimage_postimage_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_create_snapshot_with_native_type_with_base_rows_delete_preimage_postimage)
- [test_topology_upgrade_basic.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic.debug.1)
- [test_topology_upgrade_basic.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic.debug.2)
- [test_topology_upgrade_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic.debug.3)
- [test_topology_upgrade_not_stuck_after_recent_removal.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_topology_upgrade_not_stuck_after_recent_removal/Tests___Unit_Tests___test_topology_upgrade_not_stuck_after_recent_removal_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_not_stuck_after_recent_removal.debug.1)
- [test_topology_upgrade_not_stuck_after_recent_removal.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_topology_upgrade_not_stuck_after_recent_removal/Tests___Unit_Tests___test_topology_upgrade_not_stuck_after_recent_removal_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_not_stuck_after_recent_removal.debug.2)
- [test_topology_upgrade_not_stuck_after_recent_removal.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_topology_upgrade_not_stuck_after_recent_removal/Tests___Unit_Tests___test_topology_upgrade_not_stuck_after_recent_removal_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_not_stuck_after_recent_removal.debug.3)
- [test_view_build_status_migration_to_v2.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_view_build_status/Tests___Unit_Tests___test_view_build_status_migration_to_v2_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_migration_to_v2.debug.1)
- [test_view_build_status_migration_to_v2.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17625/testReport/junit/cluster/test_view_build_status/Tests___Unit_Tests___test_view_build_status_migration_to_v2_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_migration_to_v2.debug.2)
#### Build Details:
- Duration: 5 hr 32 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17636/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17636/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17636/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16740/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8658/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5485/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17636/)
#### Build Details:
- Duration: 7 hr 41 min
- Builder: spider1.cloudius-systems.com

avikivity commented on db/system_keyspace.cc:
```diff
@@ -96,6 +96,9 @@ namespace {
             system_keyspace::VIEW_BUILD_STATUS_V2,
+            system_keyspace::CDC_STREAMS_STATE,
+            system_keyspace::CDC_STREAMS_HISTORY,
+            system_keyspace::CDC_PENDING_STREAMS,
```
> cdc: add internal tables for cdc with tablets

How internal are they? Do the drivers access them? Or is the driver protocol unchanged?

avikivity commented on cdc/generation.hh:
```diff
@@ -118,6 +118,23 @@ public:
     {}
 };
 
+enum class stream_kind : int8_t {
```
"kind" is not a good fit here, they all have the same kind, maybe "disposition",

avikivity commented on cdc/metadata.cc:
```diff
@@ -203,3 +205,112 @@ bool cdc::metadata::prepare(db_clock::time_point tp) {
+                std::vector<stream_id> next_stream_set;
+                next_stream_set.reserve(prev_stream_set.size() + opened.size() - closed.size());
+
+                std::ranges::sort(closed, stream_cmp);
```
Another likely stall.

avikivity commented on cdc/log.cc:
```diff
@@ -1508,6 +1508,8 @@ class transformer final : public change_processor {
     std::vector<mutation> _result_mutations;
     std::optional<log_mutation_builder> _builder;
 
```
> cdc: double write to the pending stream

Why do we need to double-write? Can't the next generation inherit one steam during a split? And similarly one stream and adopt another during a merge?

/cc @tgrabiec

avikivity replied:
@piodul @mlitvk please address the alternator streams problem

piodul replied:
I'm looking into it. It would be nice to have it for the users of CDC, but I'm not sure yet what would be the implications for Alternator. I will probably write down the conclusions in a longer form document.

piodul replied:
https://scylladb.atlassian.net/wiki/spaces/RND/pages/137887775/CDC+for+Tablets+-+stream+switching+algorithm

The document presents two options, but I think that the most practical option at the moment would be to go with the legacy algorithm (i.e. create new streams with a timestamp that is in the future). I don't see an obvious way to adjust new algorithm to work with Alternator because of the constraints imposed by DynamoDB API.

mlitvk replied:
changed to the stream switching scheme without double writes. should work better with the alternator streams api

avikivity commented on cdc/generation.cc:
```diff
@@ -1290,6 +1290,19 @@ mutation get_open_and_close_streams_mutation(table_id table, const cdc_stream_di
     return m;
 }
 
```
> When a table's tablets are resized, if it has CDC streams, we create new

Do you mean split/merge? The tablets are not resized. The tablet count changes.

mlitvk replied:
yes. it's called resize in the code

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17671/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17671/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17671/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: cluster/test_cdc_with_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5511/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8686/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16766/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17671/)
#### Build Details:
- Duration: 8 hr 41 min
- Builder: i-01fb5db51defb4eab (m5d.16xlarge)

mlitvk commented on db/system_keyspace.cc:
```diff
@@ -337,6 +343,45 @@ schema_ptr system_keyspace::cdc_generations_v3() {
+            .with_column("stream_id", bytes_type, column_kind::clustering_key)
+            .with_column("timestamp", timeuuid_type, column_kind::static_column)
+            .with_hash_version()
+            .build();
```
it's the first timestamp and stream set for each table, that the following timestamps build upon.
when the table is created we create the first timestamp and stream set and write it to this table. when streams are changed it is written as a diff from the previous one into the cdc_streams_history table.
we can also do "compaction" when old streams are not needed anymore - we choose a new "base" timestamp and write it to cdc_streams_state with its stream set, and remove old entries from cdc_streams_history.

mlitvk commented on db/system_keyspace.cc:
```diff
@@ -337,6 +343,45 @@ schema_ptr system_keyspace::cdc_generations_v3() {
+    thread_local auto schema = [] {
+        auto id = generate_legacy_id(NAME, CDC_STREAMS_STATE);
+        return schema_builder(NAME, CDC_STREAMS_STATE, {id})
+            .with_column("table_id", uuid_type, column_kind::partition_key)
+            .with_column("stream_id", bytes_type, column_kind::clustering_key)
+            .with_column("timestamp", timeuuid_type, column_kind::static_column)
```
another thing is that currently on tablet split / merge we replace all streams at once, and not incrementally as in the design.
if that's the case, maybe we don't need to store the stream list at all? we could construct the stream id as a deterministic function of the timestamp and tablet id

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | i-07c7093521e9a3e6f |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/)<br> :small_blue_diamond: test/boost/cdc_test.cc<br> :small_blue_diamond: cluster/test_cdc_with_tablets | :white_check_mark: | spider2.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/)<br> :small_blue_diamond: test/boost/cdc_test.cc<br> :small_blue_diamond: cluster/test_cdc_with_tablets | :white_check_mark: | spider4.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17947/)<br> :small_blue_diamond: test/boost/cdc_test.cc<br> :small_blue_diamond: cluster/test_cdc_with_tablets | :white_check_mark: | spider5.cloudius-systems.com |
| dtest | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8938/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17024/) | :white_check_mark: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5746/) | :white_check_mark: |  |

#### Build Details:
- Duration: 2 hr 51 min

mlitvk commented:
I addressed some comments and integrated the PR with co-located tablets, so the CDC tablets are now co-located with the base tablets.
For that I did some changes about the create table notifications - I added a new notification pre_create_column_families and moved the CDC table creation there, and then in before_create_column_families we allocate the tablets and streams.

mlitvk commented on cdc/generation.cc:
```diff
@@ -1157,4 +1157,39 @@ db_clock::time_point get_ts(const generation_id& gen_id) {
+    m.partition().apply(tombstone(tombstone_ts, gc_now));
+    m.set_static_cell("timestamp", stream_ts, ts);
+
+    for (auto& sid : stream_ids) {
```
I added yields here and in few other places

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-063ba5179045d84a3 |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-0428fc86c7bb770fa |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-0428fc86c7bb770fa |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/)<br> :small_blue_diamond: test/boost/cdc_test.cc<br> :small_blue_diamond: cluster/test_cdc_with_tablets | :white_check_mark: | i-0428fc86c7bb770fa |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-07661cdcb59368497 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-07661cdcb59368497 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/)<br> :small_blue_diamond: test/boost/cdc_test.cc<br> :small_blue_diamond: cluster/test_cdc_with_tablets | :white_check_mark: | i-07661cdcb59368497 |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-0fdfb41ad6fa360fe |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/) | :white_check_mark: | i-0fdfb41ad6fa360fe |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18094/)<br> :small_blue_diamond: test/boost/cdc_test.cc<br> :small_blue_diamond: cluster/test_cdc_with_tablets | :white_check_mark: | i-0fdfb41ad6fa360fe |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5857/) | :x: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17169/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9084/) | :white_check_mark: |  |

#### Failed Tests (1/17400):
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18094/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/release___dtest_with_tablets___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)

#### Build Details:
- Duration: 6 hr 6 min

mlitvk commented on cdc/metadata.cc:
```diff
@@ -143,6 +158,59 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
 
+const std::vector<cdc::stream_id>& cdc::metadata::get_tablet_stream_set(table_id tid, api::timestamp_type ts) const {
+    auto now = api::new_timestamp();
+    if (ts > now + get_generation_leeway().count()) {
```
this pattern is used already in few places in cdc, and `get_generation_leeway` is defined using the type `api::timestamp_clock::duration` so it's consistent with the api timestamp, while I'm concerned that introducing `1us` explicitly whenever we use it could cause problems if something is changed

mlitvk commented on cdc/metadata.cc:
```diff
@@ -203,3 +205,112 @@ bool cdc::metadata::prepare(db_clock::time_point tp) {
+        };
+
+        auto base_stream_set = base_table_state.streams;
+        std::ranges::sort(base_stream_set, stream_cmp);
```
I added a new clustering column to the internal cdc tables with the token and moved the stream_id to be a regular column, and removed the `sort`s when loading the streams metadata from the tables.
currently in a separate commit because I wasn't sure about the idea

mlitvk commented on service/migration_listener.hh:
```diff
@@ -31,6 +31,7 @@ class function_name;
 }
 namespace locator {
 class tablet_metadata_change_hint;
+class tablet_map;
```
I squashed the commit with the commit that uses it to create streams on tablet allocation

mlitvk commented on cdc/metadata.cc:
```diff
@@ -203,3 +205,132 @@ bool cdc::metadata::prepare(db_clock::time_point tp) {
+        std::vector<cdc::stream_id> opened,
+        const std::vector<cdc::stream_id>& closed) {
+
+    if (closed.size() == prev_stream_set.size()) {
+        // all previous streams are closed, so the next stream set is just the opened streams.
+        co_return std::move(opened);
+    }
```
yes, I tried to cover it with some unit tests but not sure if it's enough coverage.
I thought maybe we can add for testing a special mode that will do asynchronous stream splitting. but maybe it's too much for now

piodul replied:
I'm not sure what you mean by asynchronous stream splitting. By that, do you mean the algorithm described in section "Deciding when to change streams" which selects only a subset of streams to change? If so, then I think it will depend on whether we go with the double writing algorithm, for now I think the current one that you implemented is OK.

mlitvk commented on service/migration_listener.hh:
```diff
@@ -80,6 +80,7 @@ public:
     // of the column family's keyspace. The reason for this is that we sometimes create a keyspace
     // and its column families together. Therefore, listeners can't load the keyspace from the
     // database. Instead, they should use the `ksm` parameter if needed.
+    virtual void on_pre_create_column_families(const keyspace_metadata& ksm, std::vector<schema_ptr>&) {}
```
how about if we rename `on_before_create_column_families` to `on_create_column_families`

mlitvk commented on service/tablet_allocator.cc:
```diff
@@ -3229,19 +3229,29 @@ class tablet_allocator_impl : public tablet_allocator::impl
+                    }
+                };
+
+                if (tm->tablets().has_tablet_map(base_id)) {
```
it replaces the previous check `if (auto it = new_cfms_map.find(base_id); it != new_cfms_map.end())`
both should be equivalent - they check if the base table is created in this operation or is this an existing table.
I used `has_tablet_map` because the next thing I'm doing is `get_tablet_map`

piodul commented on db/system_keyspace.cc:
```diff
@@ -2354,6 +2354,67 @@ future<bool> system_keyspace::cdc_is_rewritten() {
+        } else if (stream_kind == cdc::stream_kind::opened) {
+            st[tid].back().second.opened_streams.push_back(std::move(stream_id));
+        } else {
+            throw std::runtime_error(fmt::format("unexpected CDC stream kind {} in {}.{} for table {}",
+                    std::to_underlying(stream_kind), NAME, CDC_STREAMS_HISTORY, table));
```
In release mode, `on_internal_error` throws an exception instead of aborting (unless you set the `abort_on_internal_error` config option, which AFAIK we do in SCT but I doubt anybody would want to do it in production). So, in practice, `on_internal_error` will have the same effect as what you are doing right now but will result in a much more visible failure during testing.

mlitvk commented on cdc/generation.cc:
```diff
@@ -13,6 +13,7 @@
 #include <seastar/core/sleep.hh>
 #include <seastar/core/coroutine.hh>
 #include <seastar/coroutine/maybe_yield.hh>
+#include <seastar/coroutine/all.hh>
```
currently I store all the metadata in memory
I can change it to load only generations within the leeway time, and change the virtual tables to load everything

mlitvk commented on cdc/metadata.cc:
```diff
@@ -143,6 +158,57 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
+    // back until we find matching entry.
+    it = std::prev(it);
+
+    while (true) {
```
done

mlitvk commented on service/migration_manager.cc:
```diff
@@ -646,6 +653,15 @@ static future<utils::chunked_vector<mutation>> include_keyspace(
+    // We want to do this before calling `before_create_column_families`,
+    // because in `before_create_column_families` we want the subscriber to get
+    // the final list of tables.
+    db.get_notifier().pre_create_column_families(ksm, cfms);
```
done, moved this to inside the `seastar::async`

mlitvk commented on service/migration_manager.cc:
```diff
@@ -646,6 +653,15 @@ static future<utils::chunked_vector<mutation>> include_keyspace(
+    // can add a new table here (e.g. CDC).
+    // We want to do this before calling `before_create_column_families`,
+    // because in `before_create_column_families` we want the subscriber to get
+    // the final list of tables.
```
done

mlitvk commented on cdc/generation.hh:
```diff
@@ -123,6 +123,28 @@ public:
+
+using base_streams_state = std::unordered_map<table_id, committed_stream_set>;
+using streams_history = std::unordered_map<table_id, std::vector<std::pair<db_clock::time_point, cdc_stream_diff>>>;
+using pending_streams = std::unordered_map<table_id, std::vector<cdc::stream_id>>;
```
I did some changes that should help with this. I changed it to read incrementally from the tables using callbacks and not construct the full objects

avikivity commented on cdc/generation.cc:
```diff
@@ -1157,4 +1157,22 @@ db_clock::time_point get_ts(const generation_id& gen_id) {
     return std::visit([] (auto& id) { return id.ts; }, gen_id);
 }
 
```
> When allocating tablets for a CDC table, create the initial CDC stream
> set. We create one stream per each tablet, each stream covering the
> corresponding token range.

I think we should create one stream per shard that has a base table tablet. It means stream updates every time a tablet is migrated (so quite often), but fewer streams overall (100x fewer in the extreme case of a single table).

We can update it post-merge if we decide it's better.

avikivity commented on db/virtual_tables.cc:
```diff
@@ -1155,6 +1158,163 @@ class load_per_node : public group0_virtual_table {
+        return schema_builder(system_keyspace::NAME, "cdc_timestamps", std::make_optional(id))
+            .with_column("keyspace_name", utf8_type, column_kind::partition_key)
+            .with_column("table_name", utf8_type, column_kind::partition_key)
+            .with_column("timestamp", timestamp_type, column_kind::clustering_key)
```
Should use descending order so we can pick the recent events first, and use LIMIT.

mlitvk replied:
changed to descending order

avikivity commented on idl/storage_service.idl.hh:
```diff
@@ -57,6 +57,7 @@ struct raft_topology_cmd_result {
         success
     };
     service::raft_topology_cmd_result::command_status status;
+    std::optional<db_clock::time_point> local_time [[version 2025.4]];
```
Soon: [reflection](https://isocpp.org/files/papers/P2996R4.html) [[=version("2025.4")]].

avikivity commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -215,3 +226,95 @@ Note that after upgrading the cluster to 4.4, all new generations (which are cre
+^^^^^^^^^^^^^^^^^^^^^^^
+
+To list all available CDC streams for a tablets-based keyspace:
+
```
Should start by explaining the schema and columns, not dive in directly to queries.

avikivity commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -215,3 +226,95 @@ Note that after upgrading the cluster to 4.4, all new generations (which are cre
+
+      (4 rows)
+
+   Here, ``stream_kind = 1`` means that the stream was closed at this timestamp, and ``stream_kind = 2`` means that the stream was opened at this timestamp. ``stream_kind = 0`` means that the stream is active at this timestamp.
```
stream_kind should be explained more formally near the schema (and renamed!)

mlitvk replied:
added explanation about stream_state

avikivity commented:
> Looks good. Main comment is that I think we should have a stream per shard rather than stream per tablet (and open/close streams on migration rather than on split/merge). But I think this can be merged later, it doesn't change the architecture.
> 
> I'd like a report in the cover letter on an S-C-T run that includes CDC. Doesn't have to be anything special, just get some soak time on a larger configuration than we have in unit tests.

Looking at DynamoDB, they also follow one stream per partition (= our tablet), so stream per tablet isn't too onerous.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -346,6 +352,48 @@ schema_ptr system_keyspace::cdc_generations_v3() {
+        return schema_builder(NAME, CDC_STREAMS_STATE, {id})
+            .with_column("table_id", uuid_type, column_kind::partition_key)
+            .with_column("range_end", long_type, column_kind::clustering_key)
+            .with_column("stream_id", bytes_type)
```
ok

piodul commented:
@avikivity I don't think that switching to one-stream-per-shard model is that simple. The big question it raises is into which streams we should put the CDC log records.

For the sake of the example, a "replica" will mean a replica node + shard combination, and S(replica) means the stream of that replica.

Let's consider a write to the base table, targetting replicas A, B, C.

- One option is to write the CDC log record to S(A), S(B), S(C). In that case we have the same availability guarantees, but now clients need to deal with duplicates. Having to deduplicate the records would prevent parallelization of data processing from a single CDC log table (right now it's rather simple to divide the streams between multiple instances of a consumer application).
- Another option is to write the CDC log record to one, deterministically chosen stream (e.g. S(A)). This solves the duplication problem but introduces an availability problem.
- Another option, which goes in a bit different direction, would be just to have less streams (and tablets) in the CDC log table than there are tablets in the base table. This sacrifices the flexibility for distributing the data in the CDC log, and forgoes the colocation property, which in turn also diminishes availability (unless we use _hints_ to replay the log writes if log replicas are unavailable... ugh).

I think that attacking this issue would require us to reconsider some of the guarantees for CDC, which would be a much bigger decision then adjusting the current design to work with tablets.

mlitvk commented on cdc/generation.hh:
```diff
@@ -122,6 +122,27 @@ public:
+enum class stream_kind : int8_t {
+    current = 0,
+    closed = 1,
+    opened = 2,
```
added documentation in docs/dev/cdc

piodul reviewed (commented):
I reviewed the whole thing. The most important points are:

- Need to decide whether to get CDC streams GC/compaction in with this PR; if yes, then documentation should be improved
- Found one bug related to disabling CDC and re-creating the log table

Other than that and some non-nit comments, LGTM.

mlitvk commented on cdc/generation.cc:
```diff
@@ -1209,6 +1209,23 @@ future<mutation> create_table_streams_mutation(table_id table, db_clock::time_po
     co_return std::move(m);
 }
 
+future<mutation> create_table_streams_mutation(table_id table, db_clock::time_point stream_ts, const std::vector<cdc::stream_id>& stream_ids, api::timestamp_type ts) {
```
the GC uses the configured ttl of the cdc table so it should only removes information about streams for which the log rows are removed.
I think that information about these streams is not helpful because they don't have any log rows anyway

mlitvk replied:
removed the GC from this PR - will post it separately for discussion

mlitvk commented on test/cluster/test_cdc_with_tablets.py:
```diff
@@ -0,0 +1,452 @@
+                assert partitions_set.issubset(inaccessible_partitions), \
+                    f"Stream {stream_id} partitions {partitions} are not fully contained in inaccessible partitions"
+
+        await manager.server_start(servers[0].server_id)
```
yes because it's a 2 node cluster and we need quorum to drop the keyspace

mykaul commented:
@mlitvk , @piodul - this is a long series of patches - if some are reviewed and approved, let's split them to their own PRs and get the merged sooner rather than later.

mlitvk commented on db/system_keyspace.cc:
```diff
@@ -347,6 +351,35 @@ schema_ptr system_keyspace::cdc_generations_v3() {
+        auto id = generate_legacy_id(NAME, CDC_STREAMS_STATE);
+        return schema_builder(NAME, CDC_STREAMS_STATE, {id})
+            .with_column("table_id", uuid_type, column_kind::partition_key)
+            .with_column("range_end", long_type, column_kind::clustering_key)
```
renamed to last_token

mlitvk commented on test/cluster/test_cdc_with_alter.py:
```diff
@@ -25,10 +25,10 @@ async def test_add_and_drop_column_with_cdc(manager: ManagerClient):
     """
 
-    servers = await manager.servers_add(3)
+    servers = await manager.servers_add(3, auto_rack_dc="dc1")
```
expanded the commit message

mlitvk commented on test/cluster/test_cdc_with_tablets.py:
```diff
@@ -0,0 +1,452 @@
+    cql = manager.get_cql()
+    async with new_test_keyspace(manager, "WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} AND tablets = {'initial': 2}") as ks1:
+        await cql.run_async(f"CREATE TABLE {ks1}.test1 (pk int PRIMARY KEY, v int) WITH cdc={{'enabled': true}}")
+        rows = await cql.run_async("SELECT * FROM system.cdc_streams_state")
+        assert len(rows) == 2
```
added checks that the history table is empty

mlitvk commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -215,3 +221,97 @@ Note that after upgrading the cluster to 4.4, all new generations (which are cre
+
+- ``system.cdc_streams``: For each timestamp, this table shows the set of streams operating at that timestamp, as well as the changes from the previous timestamp (such as streams being opened or closed).
+  Each row includes the stream’s state (``stream_state``), which describes whether the stream is active, opened, or closed at that timestamp.
+
```
added comments to the new tables

mlitvk commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -2,6 +2,15 @@
 Querying CDC Streams
```
restructured to have the introduction first, then stream changes section, then sections for vnode and tablets

mlitvk commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -215,3 +221,97 @@ Note that after upgrading the cluster to 4.4, all new generations (which are cre
 
 .. note::
    We highly recommend using the newest releases of our client CDC libraries (`Java CDC library <https://github.com/scylladb/scylla-cdc-java>`_, `Go CDC library <https://github.com/scylladb/scylla-cdc-go>`_, `Rust CDC library <https://github.com/scylladb/scylla-cdc-rust>`_). They take care of correctly querying the stream description tables and they handle the upgrade procedure for you.
```
changed to a section "reacting to stream changes" that's before the vnodes/tablets sections

mlitvk commented on docs/features/cdc/cdc-stream-changes.rst:
```diff
@@ -1,11 +1,16 @@
+CDC Stream Generations (vnodes-based keyspaces)
+-----------------------------------------------
+
+In a vnode-based keyspace, stream IDs will only change if you join a new node to the cluster.
```
added mention of this with a reference

mlitvk commented on cdc/generation.cc:
```diff
@@ -1385,4 +1402,89 @@ future<> generation_service::generate_tablet_resize_update(utils::chunked_vector
+            continue;
+        }
+
+        auto table_muts = co_await get_cdc_stream_compaction_mutations(table, first_nonobsolete_it->second.ts, first_nonobsolete_it->second.streams, ts);
```
added info level log

mlitvk commented on cdc/generation.cc:
```diff
@@ -1385,4 +1402,89 @@ future<> generation_service::generate_tablet_resize_update(utils::chunked_vector
+    for (const auto& [table, table_streams] : _cdc_metadata.get_all_tablet_streams()) {
+        co_await coroutine::maybe_yield();
+
+        const auto& committed_streams = (*table_streams)->committed;
```
removed the `committed` member and changed the streams map to be simply a `std::map`

mlitvk commented on db/virtual_tables.cc:
```diff
@@ -1155,6 +1158,163 @@ class load_per_node : public group0_virtual_table {
+            auto& cf = _db.find_column_family(table);
+            auto base_schema = cdc::get_base_table(_db, *cf.schema());
+            auto dk = make_partition_key(base_schema->ks_name(), base_schema->cf_name());
+            if (!this_shard_owns(dk) || !contains_key(qr.partition_range(), dk)) {
```
do you think I should change it to null sharder in case we want to change it in the future?

piodul replied:
I think we can leave it for now. Queries to virtual tables are processed locally, so I think it should be possible to easily change the sharder in the future.

mlitvk commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -103,74 +182,10 @@ Example: querying the CDC description table
-and so on. After you make sure that every node uses the new generation, you can query streams from the previous generation one last time, and then switch to querying streams from the new generation.
+.. _reacting-to-topology-changes:
 
 Differences in ScyllaDB 4.3
```
done

mlitvk commented on docs/features/cdc/cdc-log-table.rst:
```diff
@@ -81,11 +81,12 @@ The ``cdc$stream_id`` column, of type ``blob``, is the log table's partition key
 Partitions in the log table are called *streams*; within one stream, all entries are sorted according to the base table writes' timestamps, using standard clustering key properties (note that ``cdc$time``, which represents the time of the write, is the first part of the clustering key).
 
-If you want to use CDC efficiently, it's important to understand how stream IDs are managed and chosen. Consult the :doc:`./cdc-streams` document for basic definitions and properties, :doc:`./cdc-stream-generations` document to understand how streams are managed and how they change over time, and finally :doc:`./cdc-querying-streams` to learn how streams can be queried efficiently, and how to fin ...
+If you want to use CDC efficiently, it's important to understand how stream IDs are managed and chosen. Consult the :doc:`./cdc-streams` document for basic definitions and properties, :doc:`./cdc-stream-changes` document to understand how streams are managed and how they change over time, and finally :doc:`./cdc-querying-streams` to learn how streams can be queried efficiently, and how to find ou ...
```
done

mlitvk commented on docs/features/cdc/cdc-querying-streams.rst:
```diff
@@ -18,16 +18,95 @@ The recommended alternative is to query each stream separately:
+Reacting to stream changes
+--------------------------
+
+As explained in :doc:`./cdc-stream-changes`, the set of used CDC stream IDs may change due to some events. You should then query the CDC description table to read the new set of stream IDs and the corresponding timestamp.
```
done

piodul commented:
Queued
